010-server-side-rendering.md

```diff
@@ -0,0 +1,205 @@
+
+Ideally, we could build each page on-demand as it changes so that content changes would remain fast. We'd still need to run a full build for app changes, though, and incremental builds for Gatsby [are not yet shipped](https://www.gatsbyjs.org/blog/2020-01-27-announcing-gatsby-builds-and-reports/#are-builds-the-same-thing-as-incremental-builds).
+
+
+#### On-Demand Rendering
+
+Our website also often veers into "application" territory, since many of our pages have stateful interfaces (for example, the campaign page changes based on whether you're signed up or not). Even some otherwise simple "content" pages have dynamic elements, like how our [referral page](https://www.dosomething.org/us/join?user_id=5543dfd6469c64ec7d8b46b3&campaign_id=9037) is customized with the referrer's first name. We could hardly pre-render that page for every single user that might possibly refer someone!
+
+For dynamic content like this, the easiest solution is to render pages or components on-demand: each time a user visits a page, the server renders the React application to HTML and sends that as the response. The user's browser can display that markup as soon as it's downloaded & then render any interaction or navigation from there on the client-side.
+
+At first glance, we're just moving the same work from the client to the (admittedly, often faster) server. However, we've got a trick up our sleeve – caching! Once we've rendered a page once, we can store it in a public cache (on [Fastly](https://www.fastly.com) or [Zeit's CDN](https://zeit.co/smart-cdn)), making subsequent visits to that page "render" instantaneously.
+
+#### Static Rendering & On-Demand Rendering
+
+Next.js supports [opt-in static generation](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation) alongside on-demand rendered routes. This means that pages that are completely static (say, an article page) can be rendered to HTML at build-time while other pages continue to be rendered on-demand. Unfortunately, this [doesn't work with Apollo's server-side rendering](#) & so we're stuck with a choice between static generation & pre-rendering colocated queries.
+
+#### Further Rendering Optimizations
+
+However, what about things like displaying a user's personalized submission gallery? We don't want a tiny bit of personalization to prevent an entire page from being cached. In those cases, we can [skip particular GraphQL queries](https://www.apollographql.com/docs/react/performance/server-side-rendering/#skipping-queries-for-ssr) when server-rendering. In this example, that allows us to render the rest of the campaign template & any user-specific elements would simply "pop" in on the client-side!
```

Comment:
Noice 👌

Reply:
So this means we'd have to tag any queries involving user id or anything user specific to have that `ssr: false` piece in order for pages using those queries to be cache-able? I imagine we'd mostlyyyy want to be caching data we get from Contentful is that correct? I'm glad there is a way to handle this!

Reply:
Yep! I think that's exactly how we'd handle this – we'd server-render and cache anything that's generic enough that most or all users could benefit from it, and client-render anything that's specific to a particular person or would otherwise cache poorly (like reaction counters in the gallery perhaps).